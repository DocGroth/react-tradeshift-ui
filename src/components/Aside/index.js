import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Portal from 'react-portal';

const noop = () => {};

class Aside extends Component {
	constructor(props) {
		super(props);
		this.onRef = this.onRef.bind(this);
		this.onOpen = this.onOpen.bind(this);
		this.onClose = this.onClose.bind(this);

		// Sync open flag for onclose / onopen callbacks
		this.isOpen = props.isOpen;
	}

	componentWillReceiveProps(nextProps) {
		this.isOpen = nextProps.isOpen;
	}

	onClose(e) {
		this.props.onClose(e);
		return typeof this.isOpen !== 'undefined' ? !this.isOpen : undefined;
	}

	onOpen(e) {
		this.props.onOpen(e);
		return this.isOpen;
	}

	onRef(ref) {
		this.ref = ref;
		if (this.ref && this.ref.spirit) {
			this.ref.spirit.onclose = this.onClose;
			this.ref.spirit.onclosed = this.props.onClosed;
			this.ref.spirit.onopen = this.onOpen;
			this.ref.spirit.onopened = this.props.onOpened;
		}
	}

	render() {
		const busyMessage = this.props.isLoading
			? this.props.loadingMessage
			: undefined;
		const asideProps = {
			'data-ts.title': this.props.title,
			'data-ts.open': this.props.isOpen,
			'data-ts.busy': busyMessage
		};
		return (
			<Portal isOpened>
				<aside data-ts="Aside" {...asideProps} ref={this.onRef}>
					<div data-ts="Panel">
						{this.props.children}
					</div>
				</aside>
			</Portal>
		);
	}
}

Aside.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.arrayOf(PropTypes.node),
		PropTypes.node
	]),
	isLoading: PropTypes.bool,
	isOpen: PropTypes.bool,
	loadingMessage: PropTypes.string,
	onClose: PropTypes.func,
	onClosed: PropTypes.func,
	onOpen: PropTypes.func,
	onOpened: PropTypes.func,
	title: PropTypes.string
};

Aside.defaultProps = {
	children: null,
	isLoading: false,
	isOpen: undefined,
	loadingMessage: 'Loading...',
	onClose: noop,
	onClosed: noop,
	onOpen: noop,
	onOpened: noop,
	title: 'Aside'
};

export default Aside;
