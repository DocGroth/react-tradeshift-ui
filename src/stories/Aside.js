import React from 'react';
import { storiesOf, action } from '@kadira/storybook';
import { text, boolean } from '@kadira/storybook-addon-knobs';
import Aside from '../components/Aside';
import Menu from '../components/Menu';
import MenuItem from '../components/MenuItem';

const stories = storiesOf('Aside', module);

stories.addWithInfo(
	'Controlled mode',
	`In controlled mode the aside will respect the isOpen attribute at all times.
	Use the onClose callback to toggle the flag.`,
	() => (
		<Aside
			title={text('title', 'Aside demo')}
			isOpen={boolean('isOpen', false)}
			loadingMessage={text('loadingMessage', 'Loading...')}
			isLoading={boolean('isLoading', false)}
			onOpen={action('onOpen')}
			onOpened={action('onOpened')}
			onClose={action('onClose')}
			onClosed={action('onClosed')}
		>
			<Menu>
				<MenuItem>
					Menu item
				</MenuItem>
			</Menu>
		</Aside>
	),
	{ inline: true }
);

stories.addWithInfo(
	'Uncontrolled mode',
	`In uncontrolled mode the aside will open on render, and
	 close automatically when the user clicks to dismiss.`,
	() => (
		<Aside
			title={text('title', 'Aside demo')}
			loadingMessage={text('loadingMessage', 'Loading...')}
			isLoading={boolean('isLoading', false)}
			onOpen={action('onOpen')}
			onOpened={action('onOpened')}
			onClose={action('onClose')}
			onClosed={action('onClosed')}
		/>
	),
	{ inline: true }
);
