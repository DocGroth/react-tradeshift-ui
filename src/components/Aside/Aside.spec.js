import React from 'react';
import { shallow } from 'enzyme';

import Aside from './';

describe('Aside', () => {
	it('renders with default props', () => {
		const wrapper = shallow(<Aside />);
		expect(wrapper).toMatchSnapshot();
	});

	describe('spirit interaction', () => {
		it('respects isOpen in controlled mode', () => {
			const wrapper = shallow(<Aside isOpen />);

			// Mock the spirit:
			const spirit = {};
			wrapper.instance().onRef({ spirit });
			expect(spirit.onclose()).toEqual(false);

			// Set callback to close the aside
			wrapper.setProps({
				onClose() {
					wrapper.setProps({ isOpen: false });
				}
			});
			expect(spirit.onclose()).toEqual(true);
		});
	});
});
