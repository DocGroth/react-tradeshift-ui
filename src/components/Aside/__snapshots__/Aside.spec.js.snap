exports[`Aside renders with default props 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Portal
    isOpened={true}
    onClose={[Function]}
    onOpen={[Function]}
    onUpdate={[Function]}>
    <aside
        data-ts="Aside"
        data-ts.busy={undefined}
        data-ts.open={undefined}
        data-ts.title="Aside">
        <div
            data-ts="Panel" />
    </aside>
</Portal>,
  "nodes": Array [
    <Portal
      isOpened={true}
      onClose={[Function]}
      onOpen={[Function]}
      onUpdate={[Function]}>
      <aside
            data-ts="Aside"
            data-ts.busy={undefined}
            data-ts.open={undefined}
            data-ts.title="Aside">
            <div
                  data-ts="Panel" />
      </aside>
</Portal>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Aside
        isLoading={false}
        isOpen={undefined}
        loadingMessage="Loading..."
        onClose={[Function]}
        onClosed={[Function]}
        onOpen={[Function]}
        onOpened={[Function]}
        title="Aside" />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Aside {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "isOpen": undefined,
        "onClose": [Function],
        "onOpen": [Function],
        "onRef": [Function],
        "props": Object {
          "children": null,
          "isLoading": false,
          "isOpen": undefined,
          "loadingMessage": "Loading...",
          "onClose": [Function],
          "onClosed": [Function],
          "onOpen": [Function],
          "onOpened": [Function],
          "title": "Aside",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Portal
          isOpened={true}
          onClose={[Function]}
          onOpen={[Function]}
          onUpdate={[Function]}>
          <aside
                    data-ts="Aside"
                    data-ts.busy={undefined}
                    data-ts.open={undefined}
                    data-ts.title="Aside">
                    <div
                              data-ts="Panel" />
          </aside>
</Portal>,
        "_debugID": 2,
        "_renderedOutput": <Portal
          isOpened={true}
          onClose={[Function]}
          onOpen={[Function]}
          onUpdate={[Function]}>
          <aside
                    data-ts="Aside"
                    data-ts.busy={undefined}
                    data-ts.open={undefined}
                    data-ts.title="Aside">
                    <div
                              data-ts="Panel" />
          </aside>
</Portal>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Aside
    isLoading={false}
    isOpen={undefined}
    loadingMessage="Loading..."
    onClose={[Function]}
    onClosed={[Function]}
    onOpen={[Function]}
    onOpened={[Function]}
    title="Aside" />,
}
`;
